### ISSUE TITLE
Feature Request: Download files with Progress Bar

### ISSUE THREAD

**Issue by gcarreno** [wp-xyz/corona#16]
Hey @wp-xyz,

You can find a version of a Modal Form with Download progress bars here:
[TestDownloadProgressBar-v1.2.zip](https://github.com/wp-xyz/corona/files/5930623/TestDownloadProgressBar-v1.2.zip)

Please take what you want if you think it's any good.

Cheers,
Gus

**Comment by wp-xyz** [comment]
Thanks for submitting. I added a simple progressbar, not exactly like you suggested, but similar.

**Comment by gcarreno** [comment]
Hey @wp-xyz,

You are quite welcome.

Well, I went overboard because I wanted to give you options, and, like I said, you could pick whatever you wanted.

On another matter, I think we had this conversation before when I reported that the download messages would never show, while the code on the action was running and I suggested using some well placed `Application.ProcessMessages()`.

At the time you dismissed it and said they were not needed. But the fact is that your reality is in a Windows environment, but my reality is in a Linux Gnome (gtk2) environment.

In your environment, and I had to use `wine` to compare, which I didn't last time, I can see the Status Messages and the Progress Bar chugging along. But not on Linux Gnome (gtk2)!!

If you don't mind, I'll do a Pull Request with some well placed `Application.ProcessMessages()`, inside `{$IFDEF LINUX}`, and submit it to your approval.

What do you say?

Cheers,
Gus

**Comment by wp-xyz** [comment]
I just checked the current version on LMDE Linux / gtk2 and I do see the download messages. I do not see the progressbar, though, which obviously cannot be parented by a statusbar in gtk2. What a mess! I'll have to investigate.

Yes, please send a pull request so that I can have a look.

**Comment by gcarreno** [comment]
Hey @wp-xyz,

It's quite strange that you see the download messages under LMDE Linux. I could never see them until I added the `Application.ProcessMessages()`.

You can see on my example that I PROFUSELY added them through my code and even then there where 1 or 2 seconds delay from the Form Window being painted and the controls being painted. I maybe have some gremlin issues on my Ubuntu Gnome install :disappointed:. If this is the case, it's gonna be really hard to know where the issue is :disappointed:.

I've closed the PR since I got nowhere with the Progress Bar with simple solutions like the `Application.ProcessMessages()` or even `ProgressBar.BringToFront()`.

I think the `TStatusBar` needs a panel of type own draw, or the likes (my memory is fuzzy on the panel types), where on the `OnDraw` event we can manage a runtime created `TProgressBar`.

I'm really sorry I couldn't be of more help!!

Cheers,
Gus

PS: Would really appreciate if you could at least add the `.gitignore` and the `.gitattributes` files for a better repo health and multiplatform handling of Windows(CRLF) and Linus(LF) line endings.
PSS: You should also remove the `source/corona.res` from the VCS, thanks!

**Comment by gcarreno** [comment]
UFFFFF!!!

Just updated `.gitignore`, deleted `*.res` and `*.lps`, committed and pushed on 26 repositories. I realized that if I was gonna preach, I should follow my own words :smiley:

**Comment by wp-xyz** [comment]
Since the retriggerable download button cannot be solved by disabling the action I decided to follow your idea with the modal download form and implemented the threaded downloader with progressbars. Thanks for the contribution. I hope it's ok that I restructured your units although you have license texts in them.

**Comment by gcarreno** [comment]
Hey @wp-xyz ,

I think you forgot to add the new files to the repo on your last commit, eheh 😋

> Since the retriggerable download button cannot be solved by disabling the action

Are you sure? This invalidates a lot of assumptions I've had in the past. I need to investigate further. MANY THANKS for the heads up!!

> Thanks for the contribution.

You're quite welcome!!! 😀

> I hope it's ok that I restructured your units although you have license texts in them.

Please, you can remove all the headers. I inserted them as a muscle memory thing and it's overkill for the small piece of code. Just drop my name somewhere, anywhere, doesn't really matter.

Cheers,
Gus

**Comment by wp-xyz** [comment]
> I think you forgot to add the new files to the repo on your last commit, eheh 😋

Grrr... Same error again and again...

> > Since the retriggerable download button cannot be solved by disabling the action
> 
> Are you sure? 

No. When I wrote this my test program (see attachment) contained a "sleep(2000)" to simulate the download and I could do whatever I wanted - i could not prevent a second click on the disabled button to restart another "download". Then I had the idea that I should invoke Application.ProcessMessage, and I divided the "sleep(2000)" into consecutive shorter sleeps with Application.ProcessMessages calls in between. And now it works in the simplest version that I can conceive.

So, maybe I could have kept the old version. But using a threaded download "manager" opens possibilities for the future. Maybe sometime I'll extend it to use several threads, or provide the possibility to abort a download waiting forever... just an idea, no distinctive plans yet.

You have some commented error log messages in the code. I know from your previous message that you are not happy with the error handling of the downloads. Feel free to improve it. I did not focus on these things because my primary interest was in the corona topic and analyzing the data in my own way to be independent of what media tell me. 

> Please, you can remove all the headers. I inserted them as a muscle memory thing and it's overkill for the small piece of code. Just drop my name somewhere, anywhere, doesn't really matter.

Yes I removed the license header now because I don't understand these license things. You are mentioned in the cDownloadManager unit.

**Comment by wp-xyz** [comment]
> Arghhh, that's my fault, **really really sorry**!!!
> I should've mentioned that you REALLY need an `Application.ProcessMessages` right after `acTheAction.Enabled:= False` so:
> 
> ```pascal
> begin
>   acTheAction.Enabled:= False;
>   Application.ProcessMessages;
>   try
>     // the rest of the code
>   finally
>     // I put this here in case the code above generates an Exception and the button is then left disabled.
>     acTheAction.Enabled:= False;
>   end;
> end;
> ```

Of course, I did this. But the Application.ProcessMessages has already been handled when the second click occurs, it is only important for the button to be painted as disabled. It is essential there are Application.ProcessMessages calls also inside the main working part, what you call "the rest of the code". Play with my test program (ah: did I forget to upload it? Here it is again:
[click_on_disabled_button.zip](https://github.com/wp-xyz/corona/files/5960503/click_on_disabled_button.zip)



> I quite understand your point of view. I'm guessing your not a professional programmer and that makes quite a bit of sense that you wanted to focus on the data.

I am phycisist and thus give higher priority to data than to rules and standards.

**Comment by gcarreno** [comment]
Hey Werner (@wp-xyz),

I've ran your code and I see your dilemma!! Which is something I was unaware of also, so MANY thanks for the investigation.

I've also noted this:
- The `TToolButton` misbehaved with no `Application.ProcessMessages()` attached to `Action1`. Like you said.
- The `TButton` misbehaved only if it wasn't attached to `Action1`. If it was attached to `Action1`, with no `Application.ProcessMessages()`, it only processed 1 `Click` and 1 `Action1`.
- When I added another button with `Action1` but no click event, it behaved like the `TToolButton`.
- The click event on `Button1` had another call to `Application.ProcessMessages()` and it made it work.
- It's consistent, but really messed up!!

I need to further investigate this also!!

> I am physicist and thus give higher priority to data than to rules and standards.

I completely agree with your point of view of a physicist. But look at it this way:
When I follow my programming standards it's like when you follow the scientific method.
So I understand that my rules and standards are irrelevant in your field, where the scientific method is the gold rule, like I understand that the scientific method doesn't really, completely, apply to programming, even if I'm somewhat of a science geek and like to have it on the back of my mind when programming/debugging :smile:

Cheers,
Gus

**Comment by gcarreno** [comment]
Hey Werner (@wp-xyz)

YAY, YAHOOO, EURECA!!!!

I've finally found a simpler solution that does not involve calling `Application.ProcessMessages()` anymore than 2 times TOTAL:

```pascal
begin
  acTheAction.Enabled:= False;
  Application.ProcessMessages; <--- This paints the associated control disabled
  try
    // the rest of the code
    { No more need to have any Application.ProcessMessages() calls in the "do work" section, YAY}
  finally
    // I put this here in case the code above generates an Exception and the button is then left disabled.
    Application.ProcessMessages; <------ THIS IS THE SOLUTION TO ALL OUR PROBLEMS!!!
    { ^^^^ I dunno what this does, being so late in the game, but it solved the messages being duplicated. }
    acTheAction.Enabled:= False;
  end;
end;
```

I'm not gonna claim credit on this one, cuz I got here by randomly placing `Application.ProcessMessages()` all over the place until it worked!!! That's the dumbest way to figure something out, but heck we got a solution!!!

But.... I will, very vehemently, THANK YOU for the test code you provided. It's been invaluable in this messing around!!!
YOU DA MAN!!!

Cheers,
Gus

**Comment by gcarreno** [comment]
Hi Werner(@wp-xyz),

Have we solved this to the point we can close this issue?

Cheers,
Gus

**Comment by wp-xyz** [comment]
For me it is finished: there is prominent display of download progress, and there is a modal form which prevents the users from clicking anywhere during download. 

I saw that you had contact with mr-highball. But at least for the moment, I am not motivated to pull in a third-party threading library for a download running for at most one or two seconds.

**Comment by gcarreno** [comment]
Hey Werner(@wp-xyz),

> For me it is finished: [...] during download.

Totally agree, will close.

> I saw that you had contact with mr-highball.

Yes I did. I'm always looking for good Thread libraries and the fact that I used the download thing was just because it was fresh in my mind.
My goal with his lib is to use on my stuff, or learn enough from his expertize, it was never to get it on corona.
As I said, corona, you and the download stuff was very fresh in my mind and it made sense at the time to ask him for an example usage of his lib to solve a thing in my mind.

Cheers,
Gus
